### PR TITLE
fix(ui): Fix semi-broken <LazyLoad>

### PR DESCRIPTION
**Pull request by billyvg**
https://github.com/getsentry/sentry/pull/34857 caused a regression where switching routes (for views that use `useRouteContext`) causes the old route component to re-render instead of unmounting. This breaks because assumptions of route parameters existing in that component no longer exist. (e.g. a details page that expects a route parameter of ID will crash when navigating to a list page because the details page gets re-rendered with the route for the list page).

Instead of using state to attempt to manage the unmounting/mounting of code split components, we can use `<Suspense>` from Reacts core API.

### CODE REVIEW

File: static/app/components/lazyLoad.tsx [hunk 1]
```diff
@@ -39,84 +39,74 @@ type Props<C extends ComponentType> = Omit<React.ComponentProps<C>, 'route'> & {
...
+  return (
+    <ErrorBoundary>
+      <Suspense
+        fallback={
+          <LoadingContainer>
+            <LoadingIndicator />
+          </LoadingContainer>
+        }
+      >
+        <LazyComponent {...(props as React.ComponentProps<C>)} />;
```
Comment (billyvg) [inline static/app/components/lazyLoad.tsx:65]: Damnit
```suggestion
        <LazyComponent {...(props as React.ComponentProps<C>)} />
```

File: static/app/components/lazyLoad.tsx [hunk 2]
```diff
@@ -39,84 +39,74 @@ type Props<C extends ComponentType> = Omit<React.ComponentProps<C>, 'route'> & {
...
+      scope.setExtra('errorInfo', errorInfo);
+      Sentry.captureException(error);
+    });
+
+    // eslint-disable-next-line no-console
+    console.error(error);
   }
 
-  return <LazyComponent {...(props as React.ComponentProps<C>)} />;
+  fetchRetry = () => this.setState({hasError: false});
```
Comment (evanpurkhiser) [inline static/app/components/lazyLoad.tsx:95]: damn this is slightly less intuitive -- but I assume it's that re-rendering the component with Suspense will try and load it again right

File: static/app/components/lazyLoad.tsx [hunk 3]
```diff
@@ -39,84 +39,74 @@ type Props<C extends ComponentType> = Omit<React.ComponentProps<C>, 'route'> & {
...
+          <LoadingContainer>
+            <LoadingIndicator />
+          </LoadingContainer>
+        }
+      >
+        <LazyComponent {...(props as React.ComponentProps<C>)} />
+      </Suspense>
+    </ErrorBoundary>
+  );
+}
```
Comment (evanpurkhiser) [inline static/app/components/lazyLoad.tsx:69]:
```suggestion
  const loadingIndicator = (
    <LoadingContainer>
      <LoadingIndicator />
    </LoadingContainer>
  );

  return (
    <ErrorBoundary>
      <Suspense fallback={loadingIndicator}>
        <LazyComponent {...(props as React.ComponentProps<C>)} />
      </Suspense>
    </ErrorBoundary>
  );
}
```

**complete nit** so feel free to leave it as is, I find this marginally easier to read.

**Review by evanpurkhiser — APPROVED** [review]
Thanks a lot for fixing this billy.

Suspense is probably what we'll need to use to replace react-router v3's `getComponent` as well.

File: static/app/components/lazyLoad.tsx (hunk 2 above)
Comment (billyvg) [reply static/app/components/lazyLoad.tsx:95]: yeah exactly

File: static/app/components/lazyLoad.tsx (hunk 3 above)
Comment (billyvg) [reply static/app/components/lazyLoad.tsx:69]: I disagree, I don't like having to jump around in the code when reading

File: static/app/components/lazyLoad.tsx [hunk 4]
```diff
@@ -92,14 +97,15 @@ class ErrorBoundary extends Component {
     console.error(error);
   }
 
-  fetchRetry = () => this.setState({hasError: false});
+  // Reset `hasError` so that we attempt to render `this.props.children` again
```
Comment (evanpurkhiser) [inline static/app/components/lazyLoad.tsx:100]: 👍

**Review by evanpurkhiser — APPROVED** [review]

**Pull request merged**